include/tvm/runtime/packed_func.h

```diff
@@ -2129,6 +2315,42 @@ struct PackedFuncValueConverter<::tvm::runtime::String> {
   }
 };
 
+template <typename T>
+struct PackedFuncValueConverter<Array<T>> {
+  static Array<T> From(const TVMArgValue& val) {
+    auto untyped_array = val.AsObjectRef<Array<ObjectRef>>();
+
+    // Attempt to convert each item of the array into the desired
+    // type.  If the items do not require a conversion, no copies are
+    // made.
+    return untyped_array.Map([](ObjectRef item) {
+      // The TVMArgValue is intentionally defined through
```

Comment:
add quick common case short cut, and avoid arg setter conversion logic
```c++
if (auto* ptr = item->as<T::ContainerType>()) {
   return GetRef<T>(ptr);
} 
```

Reply:
the main reason is that round trip through to TVMArg and TVMRet variant have cost of switching etc, and we should optimize for most freq path

Reply:
Brainstorming, I think we could avoid the round trip through `TVMArgValue` and `TVMRetValue` by instead changing the signature in `PackedFuncValueConverter` so it could make recursive calls on the `ObjectRef` directly.

```c++
template<typename T>
struct PackedFuncValueConverter<Array<T>> {
  // Current API, value converter converts from TVMArgValue or
  // TVMRetValue.  The From() function is called whenever converting
  // to Array<T>.
  static Array<T> From(const TVMArgValue& val) { ... }
};

template<typename T>
struct PackedFuncValueConverter<Array<T>> {
  // Proposed API, value converter converts from the value contained
  // within the TVMArgValue or TVMRetValue.  The From() function is
  // called whenever converting to T, if the contained value is of an
  // appropriate type.
  //
  // Exposing the individual overloads would allow them to be called
  // without first wrapping in `TVMArgValue` or `TVMRetValue`.
  static runtime::String From(ObjectRef obj) { ... }
};



// For types that are convertible from primitives, an overload
// accepting a primitive argument can be defined.
template<>
struct PackedFuncValueConverter<String> {
  static String From(const char* prim) { ... }

  static String From(const TVMByteArray* arr) { ... }
}

// When delegating to PackedFuncValueConverter, the existence of an
// overload can be checked at C++ compile-time, so the performance
// should be equivalent to the current checks.
inline TVMArgValue::operator T() const {
  switch(type_code_) {
    case kTVMArgInt:
      // If the TVMArgValue holds an integer, and
      // PackedFuncValueConverter defines a conversion from integer to
      // T, then run the converter.
      if constexpr(std::is_invocable_v<PackedFuncValueConverter<T>::From, int64_t>) {
        return PackedFuncValueConverter<T>::From(value_.v_int64);
      }
      break;

    ...
  }

  LOG(FATAL) << "Error message here";
}
```

I think such a change to `PackedFuncValueConverter` would be best done in a follow-up PR, as this one is already on the larger side.

Reply:
let us first go with this one for now